curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-string-and-array-methods-by-building-a-music-player/653639d63a45a077333312c8.md

```diff
@@ -7,23 +7,27 @@ dashedName: step-29
 
 # --description--
 
-Before playing the song, you need to make sure it starts from the beginning. This can be achieved by the use of the `currentTime` property of the `audio` object. 
+Before playing the song, you need to make sure it starts from the beginning. This can be achieved by the use of the `currentTime` property on the `audio` object. 
 
-Add an `if` statement to check whether the `userData?.currentSong` is `null` or if `userData?.currentSong.id` is strictly not equal `song.id`. Inside `if` block, set the `currentTime` property of the `audio` object to `0`.
+Add an `if` statement to check whether the `userData?.currentSong` is falsy *OR* if `userData?.currentSong.id` is strictly not equal to `song.id`. This condition will check if no current song is playing or if the current song is different from the one that is about to be played.
 
+Inside your `if` block, set the `audio.currentTime` property to `0`.
 
 # --hints--
 
-You should create an `if` statement with the condition `userData?.currentSong === null || userData?.currentSong.id !== song.id`.
+You should create an `if` statement that checks whether `userData.currentSong` is either falsy *OR* `userData.currentSong` is not equal to `song.id`. Don't forget to use optional chaining.
 
 ```js
-assert.match(code, /if\s*\(\s*userData\?\.currentSong\s*===\s*null\s*\|\|\s*userData\?\.currentSong\.id\s*!==\s*song\.id\s*\)\s*\{\s*/)
+assert.match(code, /if\s*\(\s*(?:!userData\?\.currentSong|\s*userData\?\.currentSong\s*===\s*null|\s*userData\?\.currentSong\s*===\s*undefined)\s*\|\|\s*userData\?\.currentSong\.id\s*!==\s*song\.id\s*|\s*userData\?\.currentSong\.id\s*!==\s*song\.id\s*\|\|\s*(?:!userData\?\.currentSong|\s*userData\?\.currentSong\s*===\s*null|\s*userData\?\.currentSong\s*===\s*undefined)\s*\)\s*\{\s*/);
```

Comment:
```suggestion
assert.match(code, /if\s*\(\s*(?:(?:!userData\?\.currentSong|\s*userData\?\.currentSong\s*===\s*null|\s*userData\?\.currentSong\s*===\s*undefined)\s*\|\|\s*userData\?\.currentSong\.id\s*!==\s*song\.id\s*|\s*userData\?\.currentSong\.id\s*!==\s*song\.id\s*\|\|\s*(?:!userData\?\.currentSong|\s*userData\?\.currentSong\s*===\s*null|\s*userData\?\.currentSong\s*===\s*undefined))\s*\)\s*\{/);
```
There's missing one grouping, what allows passing with only part of code:
```javascript
if (!userData?.currentSong || userData?.currentSong.id !== song.id
```
```javascript
userData?.currentSong.id !== song.id || !userData?.currentSong ) {
```

Reply:
@gikf The code snippets you've provided work even without this suggestion. Please explain what this suggestion does.

Reply:
@aaqib605 That's the issue, they shouldn't pass, as that's invalid code.

Reply:
@gikf Got it. I made the necessary changes, and now it should work fine.

Reply:
Are you sure? It doesn't look like anything was changed...

Reply:
@gikf Could you provide a screenshot highlighting the issue with the code?

Reply:
Line 21:
https://github.com/freeCodeCamp/freeCodeCamp/blob/7d9c3be23793b7ce5090649961a967ed2daf33b4/curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-string-and-array-methods-by-building-a-music-player/653639d63a45a077333312c8.md?plain=1#L21
As in the first comment with suggestion how to change it below:
```suggestion
assert.match(code, /if\s*\(\s*(?:(?:!userData\?\.currentSong|\s*userData\?\.currentSong\s*===\s*null|\s*userData\?\.currentSong\s*===\s*undefined)\s*\|\|\s*userData\?\.currentSong\.id\s*!==\s*song\.id\s*|\s*userData\?\.currentSong\.id\s*!==\s*song\.id\s*\|\|\s*(?:!userData\?\.currentSong|\s*userData\?\.currentSong\s*===\s*null|\s*userData\?\.currentSong\s*===\s*undefined))\s*\)\s*\{/);
```

- - -
Similarly with https://github.com/freeCodeCamp/freeCodeCamp/pull/53786/files#r1499699708, no changes were made, semicolon is still required by regex.
https://github.com/freeCodeCamp/freeCodeCamp/blob/7d9c3be23793b7ce5090649961a967ed2daf33b4/curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-string-and-array-methods-by-building-a-music-player/653639d63a45a077333312c8.md?plain=1#L29